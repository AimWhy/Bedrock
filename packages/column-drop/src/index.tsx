import {
  CSSLength,
  Gutter,
  SizesOptions,
  getSafeGutter,
  getSizeValue,
  useTheme,
} from "@bedrock-layout/spacing-constants";
import { forwardRefWithAs } from "@bedrock-layout/type-utils";
import React, { CSSProperties } from "react";

type MinItemWidth = CSSLength | number | SizesOptions;

export interface ColumnDropProps {
  gutter?: Gutter;
  minItemWidth?: MinItemWidth;
  noStretchedColumns?: boolean;
}

function getSafeMinItemWidth(
  theme: {
    space?: { [key: string]: string };
    sizes?: { [key: string]: string };
  },
  minItemWidth?: MinItemWidth,
) {
  return getSizeValue(theme, minItemWidth);
}

export const ColumnDrop = forwardRefWithAs<"div", ColumnDropProps>(
  (
    { as, gutter, style, minItemWidth, noStretchedColumns = false, ...props },
    ref,
  ) => {
    const theme = useTheme();
    const maybeGutter = getSafeGutter(theme, gutter);

    const attributeValue =
      noStretchedColumns === true ? "no-stretched-columns" : "";

    const safeMinItemWidth = getSafeMinItemWidth(theme, minItemWidth);
    const safeStyle = style ?? {};

    const Component = as ?? "div";
    return (
      <Component
        ref={ref}
        data-bedrock-column-drop={attributeValue}
        style={
          {
            ...safeStyle,
            "--gutter": maybeGutter,
            "--minItemWidth": safeMinItemWidth,
          } as CSSProperties
        }
        {...props}
      />
    );
  },
);
ColumnDrop.displayName = "ColumnDrop";
