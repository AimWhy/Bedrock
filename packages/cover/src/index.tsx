import {
  CSSLength,
  Gutter,
  SizesOptions,
  getSafeGutter,
  getSizeValue,
  useTheme,
} from "@bedrock-layout/spacing-constants";
import { forwardRefWithAs } from "@bedrock-layout/type-utils";
import React, { CSSProperties } from "react";

/**
 * `minHeight` can be a CSSLength, a number, or a key of the theme's sizes object
 */
export type MinHeight = CSSLength | number | SizesOptions;

/**
 * Props for the Cover component.
 */
export type CoverProps = {
  /**
   * Slot to be placed at the top of the cover component, above the centered content.
   */
  top?: React.ReactNode;
  /**
   * Slot to be placed at the bottom of the cover component, below the centered content.
   */
  bottom?: React.ReactNode;
  /**
   * Sets space between each element.
   * @deprecated Use `gap` instead.
   */
  gutter?: Gutter;
  /**
   * Sets space between each element.
   */
  gap?: Gutter;
  /**
   * Sets the minimum height of the cover component.
   * `minHeight` can be a CSSLength, a number, or a key of the theme's sizes object
   * @default "100vh"
   */
  minHeight?: MinHeight;
  /**
   * Sets the content to stretch to the full height of the cover component minus the top and bottom slots.
   */
  stretchContent?: boolean;
};

/**
 * The `Cover` component is designed to vertically cover a predefined area, `100vh` by default, and vertically center its children.
 * You can also conditionally render a top and/or bottom slot as well.
 */
export const Cover = forwardRefWithAs<"div", CoverProps>(function Cover(
  {
    as: Component = "div",
    children,
    gap,
    gutter,
    top,
    bottom,
    minHeight,
    style = {},
    //TODO: Change from boolean to string union of "stretch-content" | "default"
    stretchContent,
    ...props
  },
  ref,
) {
  const theme = useTheme();
  const maybeGutter = getSafeGutter(theme, gap ?? gutter);
  const maybeMinHeight = getSizeValue(theme, minHeight);

  const attributeVal = stretchContent === true ? "stretch-content" : "";

  return (
    <Component
      ref={ref}
      data-bedrock-cover={attributeVal}
      style={
        {
          "--gutter": maybeGutter,
          "--minHeight": maybeMinHeight,
          ...style,
        } as CSSProperties
      }
      {...props}
    >
      {top && <div data-bedrock-cover-top="">{top}</div>}
      <div data-bedrock-cover-centered="">{children}</div>
      {bottom && <div data-bedrock-cover-bottom="">{bottom}</div>}
    </Component>
  );
});
