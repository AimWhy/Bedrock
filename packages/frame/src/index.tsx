import { forwardRefWithAs } from "@bedrock-layout/type-utils";
import React, { CSSProperties } from "react";
type RatioString = `${number}/${number}` | `${number} / ${number}`;

type Ratio = [number, number] | RatioString;
export interface FrameProps {
  ratio?: Ratio;
  position?: string;
}

function checkIsRatio(ratio: unknown): ratio is Ratio {
  const isCorrectArray =
    Array.isArray(ratio) && ratio.length === 2 && ratio.every(Number.isFinite);
  return (
    isCorrectArray ||
    (typeof ratio === "string" &&
      /^\d{1,1000} {0,1}\/ {0,1}\d{1,1000}$/.test(ratio))
  );
}

function getRatioString(ratio: Ratio): RatioString {
  return Array.isArray(ratio) ? (ratio.join("/") as RatioString) : ratio;
}

function getSafeRatio(ratio: unknown): RatioString | undefined {
  const isRatio = checkIsRatio(ratio);

  return isRatio ? getRatioString(ratio) : undefined;
}

export const Frame = forwardRefWithAs<"div", FrameProps>(
  ({ as, ratio, style, position, ...props }, ref) => {
    const safeRatio = getSafeRatio(ratio);
    const safeStyle = style ?? {};

    const Component = as ?? "div";

    return (
      <Component
        data-bedrock-frame
        ref={ref}
        style={
          {
            ...safeStyle,
            "--ratio": safeRatio,
            "--position": position,
          } as CSSProperties
        }
        {...props}
      />
    );
  }
);

Frame.displayName = "Frame";
