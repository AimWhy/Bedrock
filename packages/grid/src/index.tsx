import {
  SpacingOptions,
  getSpacingValue,
} from "@bedrock-layout/spacing-constants";
import PropTypes from "prop-types";
import styled from "styled-components";

const SMALL_BREAKPOINT = "639px";

export interface GridProps {
  gutter: keyof SpacingOptions;
  minItemWidth?: number | string;
}

function getSafeMinItemWidth(minItemWidth?: number | string) {
  if (typeof minItemWidth === "string") {
    return minItemWidth;
  }

  return typeof minItemWidth === "number"
    ? `${minItemWidth}px`
    : SMALL_BREAKPOINT;
}

export const Grid = styled.div.attrs<GridProps>(() => {
  return {
    "data-bedrock-layout-grid": "",
  };
})<GridProps>`
  box-sizing: border-box;
  @property --minItemWidth {
    syntax: "<length-percentage>";
    inherits: false;
    initial-value: ${SMALL_BREAKPOINT};
  }

  --gutter: ${({ gutter, theme }) => getSpacingValue(theme, gutter) ?? "0px"};
  --minItemWidth: ${({ minItemWidth }) => getSafeMinItemWidth(minItemWidth)};

  display: grid;
  gap: var(--gutter);

  grid-template-columns: repeat(
    auto-fit,
    minmax(min(${SMALL_BREAKPOINT}, 100%), 1fr)
  );

  @supports (
    width: ${({ minItemWidth }) => getSafeMinItemWidth(minItemWidth)}
  ) {
    grid-template-columns: repeat(
      auto-fit,
      minmax(min(var(--minItemWidth), 100%), 1fr)
    );
  }
`;

Grid.displayName = "Grid";

Grid.propTypes = {
  gutter: PropTypes.string.isRequired as React.Validator<keyof SpacingOptions>,
  minItemWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};
