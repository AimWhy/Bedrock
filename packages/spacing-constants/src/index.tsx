import OpenProps from "open-props";
import React, { createContext, useContext } from "react";

type Maybe<T> = NonNullable<T> | undefined;

function convertToMaybe<T extends unknown>(value: T): Maybe<T> {
  return value ?? undefined;
}

// eslint-disable-next-line @typescript-eslint/ban-types
export type DefaultTheme = {};

export const ThemeContext = createContext<DefaultTheme>({});

export function useTheme(): DefaultTheme {
  return useContext(ThemeContext);
}

export function ThemeProvider({
  theme,
  children,
}: {
  theme: DefaultTheme;
  children: React.ReactNode;
}): JSX.Element {
  return (
    <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
  );
}

type LowercaseCharacter =
  | "a"
  | "b"
  | "c"
  | "d"
  | "e"
  | "f"
  | "g"
  | "h"
  | "i"
  | "j"
  | "k"
  | "l"
  | "m"
  | "n"
  | "o"
  | "p"
  | "q"
  | "r"
  | "s"
  | "t"
  | "u"
  | "v"
  | "w"
  | "x"
  | "y"
  | "z";
type AllCharacter = LowercaseCharacter | Uppercase<LowercaseCharacter>;
type NonEmptyString = `${AllCharacter}${string}`;

type CSSCustomProperties = `var(--${NonEmptyString})`;

type LengthUnit =
  | "vmin"
  | "vmax"
  | "vh"
  | "vw"
  | "%"
  | "ch"
  | "ex"
  | "em"
  | "rem"
  | "in"
  | "cm"
  | "mm"
  | "pt"
  | "pc"
  | "px";

export type CSSLength = `${number}${LengthUnit}` | CSSCustomProperties;

function fromEntries<K extends string | number, T>(
  entries: [s: K, value: T][]
): Record<K, T> {
  return entries.reduce((acc, [key, value]) => {
    return { ...acc, [key]: value };
  }, {} as Record<K, T>);
}

const {
  sizeXxs,
  sizeXs,
  sizeSm,
  sizeMd,
  sizeLg,
  sizeXl,
  sizeXxl,
  sizeContent1,
  sizeContent2,
  sizeContent3,
  sizeHeader1,
  sizeHeader2,
  sizeHeader3,
  size000,
  size00,
  size1,
  size2,
  size3,
  size4,
  size5,
  size6,
  size7,
  size8,
  size9,
  size10,
  size11,
  size12,
  size13,
  size14,
  size15,
} = OpenProps;

const space = {
  size000,
  size00,
  size1,
  size2,
  size3,
  size4,
  size5,
  size6,
  size7,
  size8,
  size9,
  size10,
  size11,
  size12,
  size13,
  size14,
  size15,
};

const size = {
  sizeXxs,
  sizeXs,
  sizeSm,
  sizeMd,
  sizeLg,
  sizeXl,
  sizeXxl,
  sizeContent1,
  sizeContent2,
  sizeContent3,
  sizeHeader1,
  sizeHeader2,
  sizeHeader3,
};

export const spacing = space;

export const sizes = size;

export function checkIsCSSLength(str: string): str is CSSLength {
  if (typeof str !== "string") return false;

  return [
    /^[0-9]{0,10000}\.?[0-9]{1,10000}(vmin|vmax|vh|vw|%|ch|ex|em|rem|in|cm|mm|pt|pc|px)$/,
    /^var\(--\D{1,100}\)$/,
  ].some((regex) => regex.test(str));
}

export type BaseTheme = Record<string, CSSLength | string | number>;

type ThemeOrDefaultSpace<T> = T extends {
  space: BaseTheme;
}
  ? T["space"]
  : keyof typeof spacing;

type ThemeOrDefaultSizes<T> = T extends {
  sizes: BaseTheme;
}
  ? T["sizes"]
  : keyof typeof sizes;

export type SpacingOptions = ThemeOrDefaultSpace<DefaultTheme>;
export type SizesOptions = ThemeOrDefaultSizes<DefaultTheme>;

export function getSpacingValue<T>(
  theme: T & {
    spacing?: BaseTheme;
    space?: BaseTheme;
  },
  spacingKey: SpacingOptions
): Maybe<CSSLength> {
  const maybeSpacingOrDefault = theme.space ?? spacing;

  const safeSpacings = fromEntries(
    Object.entries(maybeSpacingOrDefault).map(([spaceKey, value]) => [
      spaceKey as SpacingOptions,
      (typeof value === "number" ? `${value}px` : value) as CSSLength,
    ])
  );

  return convertToMaybe(safeSpacings[spacingKey]);
}

export type Gutter = CSSLength | number | SpacingOptions;

export function getSafeGutter<T extends BaseTheme>(
  theme: T,
  gutter?: Gutter
): Maybe<CSSLength> {
  if (typeof gutter === "number" && gutter > 0) return `${gutter}px`;

  const isCSSLength = checkIsCSSLength(gutter as string);
  if (isCSSLength) return gutter as CSSLength;

  return gutter !== undefined
    ? getSpacingValue(theme, gutter as SpacingOptions)
    : undefined;
}

export function getSizeValue<T>(
  theme: T & {
    sizes?: BaseTheme;
  },
  sizeKey: SizesOptions
): Maybe<CSSLength> {
  if (typeof sizeKey === "number" || sizeKey === undefined) return undefined;

  const maybeSizesOrDefault = theme.sizes ?? sizes;

  const safeSizes = fromEntries(
    Object.entries(maybeSizesOrDefault).map(([sizeKey, value]) => [
      sizeKey as SizesOptions,
      (typeof value === "number" ? `${value}px` : value) as CSSLength,
    ])
  );

  return convertToMaybe(safeSizes[sizeKey]);
}
