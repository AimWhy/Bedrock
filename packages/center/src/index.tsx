import { mergeBreakpoints } from "@bedrock-layout/spacing-constants";
import PropTypes from "prop-types";
import styled from "styled-components";

export interface CenterProps {
  maxWidth?: number | string;
  centerText?: boolean;
  centerChildren?: boolean;
}

function getSafeMaxWidth(
  breakPoints: Record<string, unknown>,
  maxWidth?: number | string
) {
  if (typeof maxWidth === "string") {
    return maxWidth;
  }

  return typeof maxWidth === "number"
    ? `${maxWidth}px`
    : mergeBreakpoints(breakPoints).medium + "px";
}

export const Center = styled.div.attrs<CenterProps>(() => {
  return {
    "data-bedrock-layout-center": "",
  };
})<CenterProps>`
  @property --maxWidth {
    syntax: "<length>";
    inherits: false;
    initial-value: 1023px;
  }
  --maxWidth: ${({ maxWidth, theme: { breakPoints } }) =>
    getSafeMaxWidth(breakPoints, maxWidth)};

  box-sizing: content-box;

  margin-inline-start: auto;
  margin-inline-end: auto;
  margin-inline: auto;

  max-inline-size: 1023px;

  @supports (
    max-inline-size:
      ${({ maxWidth, theme: { breakPoints } }) =>
        getSafeMaxWidth(breakPoints, maxWidth)}
  ) {
    max-inline-size: var(--maxWidth, 1023px);
  }

  ${(props) => (props.centerText ? "text-align: center;" : "")}

  ${(props) =>
    props.centerChildren
      ? `
  display: flex;
  flex-direction: column;
  align-items: center;
  `
      : ""}
`;

Center.displayName = "Center";

Center.propTypes = {
  maxWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  centerText: PropTypes.bool,
};
