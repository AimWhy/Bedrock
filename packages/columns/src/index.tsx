import { Gutter, getSafeGutter } from "@bedrock-layout/spacing-constants";
import { Stack, StackProps } from "@bedrock-layout/stack";
import {
  PolymorphicComponentPropsWithRef,
  PolymorphicRef,
} from "@bedrock-layout/type-utils";
import { useContainerQuery } from "@bedrock-layout/use-container-query";
import React, { ElementType, forwardRef } from "react";
import styled from "styled-components";

interface ColumnsBaseProps {
  gutter?: Gutter;
  columns?: number;
  dense?: boolean;
  forwardedAs?: unknown;
}

const ColumnsBase = styled.div.attrs<ColumnsBaseProps>(
  ({ dense, theme, gutter, columns = 1, style }) => {
    const maybeGutter = getSafeGutter(theme, gutter);
    const safeColumns = columns > 0 ? columns : 1;
    return {
      "data-bedrock-columns": dense ? "dense" : "",
      style: { ...style, "--gutter": maybeGutter, "--columns": safeColumns },
    };
  }
)<ColumnsBaseProps>`
  @property --gutter {
    syntax: "<length-percentage>";
    inherits: false;
    initial-value: 0;
  }

  @property --columns {
    syntax: "<number>";
    inherits: true;
    initial-value: 1;
  }

  box-sizing: border-box;
  > * {
    margin: 0;
  }

  display: grid;
  grid-template-columns: repeat(var(--columns, 1), 1fr);
  gap: var(--gutter, 0px);
  grid-auto-flow: row ${({ dense = false }) => (dense === true ? "dense" : "")};
`;

interface ColumnsPropsBase extends StackProps, ColumnsBaseProps {
  switchAt?: number | string;
  children?: React.ReactNode;
}

export type ColumnsProps<C extends ElementType = "div"> =
  PolymorphicComponentPropsWithRef<C, ColumnsPropsBase>;

const ColumnComp = forwardRef(
  <C extends ElementType = "div">(
    { columns, dense, switchAt, as, gutter, ...props }: ColumnsProps<C>,
    ref?: PolymorphicRef<C>
  ) => {
    const maybePx = React.useMemo(() => {
      return typeof switchAt === "string"
        ? toPX(switchAt)
        : typeof switchAt === "number" && switchAt > -1
        ? switchAt
        : null;
    }, [switchAt]);

    const widthToSwitchAt: number = maybePx ?? 0; //zero is used to make the switchAt a noop

    const [shouldSwitch, safeRef] = useContainerQuery(
      { width: widthToSwitchAt },
      ref
    );

    return shouldSwitch ? (
      <Stack as={as} ref={safeRef} gutter={gutter} {...props} />
    ) : (
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      //@ts-expect-error
      <ColumnsBase
        as={as}
        ref={safeRef}
        gutter={gutter}
        columns={columns}
        dense={dense}
        {...props}
      />
    );
  }
);

export const Columns = styled(ColumnComp).attrs(({ as, forwardedAs }) => {
  return {
    forwardedAs: as ?? forwardedAs,
    as: ColumnComp,
  };
})``;

Columns.displayName = "Columns";

export interface ColumnProps {
  span?: number;
  offsetStart?: number;
  offsetEnd?: number;
}

const safeSpan = (span: unknown) => {
  return typeof span === "number" ? span : 1;
};

/**
 *
 * span is remaped to colSpan due to span being an attribute that gets
 * passed to the underlying element.  This can cause issues with Grid layout.
 *
 * In a future breaking change, colSpan should be the public API.
 * */
export const Column = styled.div.attrs<ColumnProps>(
  ({ span, style, offsetStart = 0, offsetEnd = 0 }) => {
    const safeOffsetStart = offsetStart > 0 ? offsetStart : undefined;
    const safeOffsetEnd = offsetEnd > 0 ? offsetEnd : undefined;
    return {
      "data-bedrock-column": "",
      span: undefined,
      style: {
        ...style,
        "--span": Math.max(safeSpan(span), 1),
        "--offsetStart": safeOffsetStart,
        "--offsetEnd": safeOffsetEnd,
      },
    };
  }
)<ColumnProps>`
  @property --span {
    syntax: "<number>";
    inherits: true;
    initial-value: 1;
  }

  @property --offsetStart {
    syntax: "<number>";
    inherits: true;
    initial-value: 1;
  }

  @property --offsetEnd {
    syntax: "<number>";
    inherits: true;
    initial-value: 1;
  }

  grid-column: span min(var(--span, 1), var(--columns, 1));

  &[style*="--offset"] {
    display: contents;
  }

  &[style*="--offset"] > * {
    grid-column: span min(var(--span, 1), var(--columns, 1));
  }

  &[style*="--offsetStart"]::before {
    content: "";
    grid-column: span min(var(--offsetStart, 1), var(--columns, 1));
  }

  &[style*="--offsetEnd"]::after {
    content: "";
    grid-column: span min(var(--offsetEnd, 1), var(--columns, 1));
  }
`;

Column.displayName = "Column";

/**
 * This module is adapted from https://github.com/mikolalysenko/to-px/blob/master/browser.js
 */

const PIXELS_PER_INCH = 96;

/* c8 ignore start */
function parseUnit(str: string): [number, string] {
  str = String(str);
  const num = parseFloat(str);

  /* c8 ignore next */
  const [, unit] = str.match(/[\d.\-+]*\s*(.*)/) ?? ["", ""];

  return [num, unit];
}

/* c8 ignore next */
function toPX(str: string, element?: Element): number | null {
  /* c8 ignore next */
  if (!str) return null;

  const elementOrBody = element ?? document.body;
  const safeStr = (str ?? "px").trim().toLowerCase();

  /* c8 ignore next */
  switch (safeStr) {
    case "vmin":
    case "vmax":
    case "vh":
    case "vw":
    case "%":
      return null;
    case "ch":
    case "ex":
      return getSizeBrutal(safeStr, elementOrBody);
    case "em":
      return getPropertyInPX(elementOrBody, "font-size");
    case "rem":
      return getPropertyInPX(document.body, "font-size");
    case "in":
      return PIXELS_PER_INCH;
    case "cm":
      return PIXELS_PER_INCH / 2.54;
    case "mm":
      return PIXELS_PER_INCH / 25.4;
    case "pt":
      return PIXELS_PER_INCH / 72;
    case "pc":
      return PIXELS_PER_INCH / 6;
    case "px":
      return 1;
    default: {
      const [value, units] = parseUnit(safeStr);

      if (isNaN(value)) return null;

      if (!units) return value;

      const px = toPX(units, element);
      return typeof px === "number" ? value * px : null;
    }
  }
}

/* c8 ignore next */
function getPropertyInPX(element: Element, prop: string): number {
  const [value, units] = parseUnit(
    getComputedStyle(element).getPropertyValue(prop)
  );
  return value * (toPX(units, element) ?? 1);
}

function getSizeBrutal(unit: string, element: Element) {
  const testDIV = document.createElement("div");
  testDIV.style["height"] = "128" + unit;
  element.appendChild(testDIV);
  const size = getPropertyInPX(testDIV, "height") / 128;
  element.removeChild(testDIV);
  return size;
}
/* c8 ignore stop */
