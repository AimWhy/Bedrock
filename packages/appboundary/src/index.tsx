import { Center } from "@bedrock-layout/center";
import { mergeBreakpoints } from "@bedrock-layout/spacing-constants";
import React from "react";
import styled, { ThemeContext } from "styled-components";

const safeTheme = { breakPoints: {} };

export const AppBoundary = styled.div.attrs((props) => {
  const { breakPoints } = React.useContext(ThemeContext) ?? safeTheme;
  const maxWidth = mergeBreakpoints(breakPoints).xxlarge;
  return {
    "data-bedrock-layout-appboundary": "",
    children: <Center maxWidth={maxWidth}>{props.children}</Center>,
  };
})`
  padding: 0;
  max-width: 100%;
  overflow: hidden;
  height: 100%;
`;

AppBoundary.displayName = "AppBoundary";
