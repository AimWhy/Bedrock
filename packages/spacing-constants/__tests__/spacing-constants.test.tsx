import { describe, expect, it, test } from "vitest";

import * as constants from "../src";

describe("Spacing Constants", () => {
  test("constant snapshot", () => {
    expect(constants).toMatchSnapshot();
  });

  it("should return a string if no spacing provided", () => {
    expect(constants.getSpacingValue({}, "size3")).toBe("1rem");
  });

  it("should return undefined if no spacing provided and incorrect key", () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    expect(constants.getSpacingValue({}, "noKey")).toBe(undefined);
  });

  it("should return a string if spacing provided", () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    expect(constants.getSpacingValue({ space: { "1x": "1rem" } }, "1x")).toBe(
      "1rem"
    );
  });

  it("should return a string if spacing provided using number", () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    expect(constants.getSpacingValue({ space: { "1x": 1 } }, "1x")).toBe("1px");
  });

  it("should return a undefined if spacing provided with incorrect key", () => {
    expect(
      constants.getSpacingValue({ space: { "1x": "1rem" } }, "size3")
    ).toBe(undefined);
  });
});

describe("gutter helpers", () => {
  describe("getSafeGutter", () => {
    const theme = { space: constants.spacing };

    it.each([
      [undefined, undefined],
      [0, "0px"],
      ["garbage", undefined],
      [{ value: "1rem" }, undefined],
      ["px", undefined],
      ["1", undefined],
      ["em1", undefined],
      ["1me", undefined],
      [" ", undefined],
      ["var(--)", undefined],
      ["var(--1)", undefined],
      [1, "1px"],
      [100, "100px"],
      ["1em", "1em"],
      ["0.25em", "0.25em"],
      ["11em", "11em"],
      ["110000000em", "110000000em"],
      ["1vmin", "1vmin"],
      ["1vmax", "1vmax"],
      ["1vh", "1vh"],
      ["1vw", "1vw"],
      ["1%", "1%"],
      ["1ch", "1ch"],
      ["1ex", "1ex"],
      ["1em", "1em"],
      ["1rem", "1rem"],
      ["1in", "1in"],
      ["1cm", "1cm"],
      ["1mm", "1mm"],
      ["1pt", "1pt"],
      ["1pc", "1pc"],
      ["1px", "1px"],
      ["var(--yellow)", "var(--yellow)"],
      ["var(--x)", "var(--x)"],
      ["size3", constants.spacing["size3"]],
      ["--x", "var(--x)"],
    ])("returns correct value", (gutter, expected) => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      expect(constants.getSafeGutter(theme, gutter)).toBe(expected);
    });
  });
});

describe("Size Constants", () => {
  it("should return a string if no sizes provided", () => {
    expect(constants.getSizeValue({}, "sizeLg")).toBe("1024px");
  });

  it("should return undefined if no sizes provided and incorrect key", () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    expect(constants.getSizeValue({}, "noKey")).toBe(undefined);
  });

  it("should return a string if sizes provided", () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    expect(constants.getSizeValue({ sizes: { "1x": "1199px" } }, "1x")).toBe(
      "1199px"
    );
  });

  it("should return a string if sizes provided using number", () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    expect(constants.getSizeValue({ sizes: { "1x": 1 } }, "1x")).toBe("1px");
  });

  it("should convert number to px if sizekey is a number", () => {
    expect(constants.getSizeValue({ sizes: { "1x": 1 } }, 1)).toBe("1px");
  });

  it("should return a undefined if sizes provided with incorrect key", () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    expect(constants.getSizeValue({ sizes: { "1x": "1199px" } }, "size3")).toBe(
      undefined
    );
  });
});

describe("checkIsCSSLength", () => {
  it.each([
    ["garbage", false],
    [{ value: "1rem" }, false],
    ["px", false],
    [1, false],
    ["1", false],
    ["em1", false],
    ["1me", false],
    [" ", false],
    ["1em", true],
    ["0.25em", true],
    ["11em", true],
    ["110000000em", true],
    ["1vmin", true],
    ["1vmax", true],
    ["1vh", true],
    ["1vw", true],
    ["1%", true],
    ["1ch", true],
    ["1ex", true],
    ["1em", true],
    ["1rem", true],
    ["1in", true],
    ["1cm", true],
    ["1mm", true],
    ["1pt", true],
    ["1pc", true],
    ["1px", true],
    ["var(--yellow)", true],
    ["var(--x)", true],
    ["var(--)", false],
    ["var(--1)", false],
    ["--size-3", true],
    ["var(--size-3)", true],
    ["var(--size-3, 100px)", true],
  ])("checkIsCSSLength(%s) -> %s", (value, expected) => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-expect-error
    expect(constants.checkIsCSSLength(value)).toBe(expected);
  });
});
