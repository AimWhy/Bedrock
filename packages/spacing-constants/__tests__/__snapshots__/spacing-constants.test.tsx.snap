// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spacing Constants > constant snapshot 1`] = `
{
  "ThemeContext": {
    "$$typeof": Symbol(react.context),
    "Consumer": {
      "$$typeof": Symbol(react.context),
      "_context": [Circular],
    },
    "Provider": {
      "$$typeof": Symbol(react.provider),
      "_context": [Circular],
    },
    "_currentRenderer": null,
    "_currentRenderer2": null,
    "_currentValue": {},
    "_currentValue2": {},
    "_defaultValue": null,
    "_globalName": null,
    "_threadCount": 0,
  },
  "ThemeProvider": [Function],
  "getSafeGutter": [Function],
  "getSizeValue": [Function],
  "getSpacingValue": [Function],
  "sizes": {
    "sizeContent1": "20ch",
    "sizeContent2": "45ch",
    "sizeContent3": "60ch",
    "sizeHeader1": "20ch",
    "sizeHeader2": "25ch",
    "sizeHeader3": "35ch",
    "sizeLg": "1024px",
    "sizeMd": "768px",
    "sizeSm": "480px",
    "sizeXl": "1440px",
    "sizeXs": "360px",
    "sizeXxl": "1920px",
    "sizeXxs": "240px",
  },
  "spacing": {
    "size00": "-.25rem",
    "size000": "-.5rem",
    "size1": ".25rem",
    "size10": "5rem",
    "size11": "7.5rem",
    "size12": "10rem",
    "size13": "15rem",
    "size14": "20rem",
    "size15": "30rem",
    "size2": ".5rem",
    "size3": "1rem",
    "size4": "1.25rem",
    "size5": "1.5rem",
    "size6": "1.75rem",
    "size7": "2rem",
    "size8": "3rem",
    "size9": "4rem",
  },
  "useTheme": [Function],
}
`;
